slice-core/src/main/java/com/cognifide/slice/core/internal/monitoring/ModelUsageData.java

```diff
@@ -0,0 +1,80 @@
+/*-
+ * #%L
+ * Slice - Core
+ * %%
+ * Copyright (C) 2012 Cognifide Limited
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.cognifide.slice.core.internal.monitoring;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import com.google.inject.Key;
+
+public class ModelUsageData {
+
+	private ConcurrentHashMap<Class<?>, ModelUsageData> subModels = new ConcurrentHashMap<Class<?>, ModelUsageData>();
+
+	private AtomicLong count = new AtomicLong();
+
+	private AtomicLong totalTime = new AtomicLong();
+
+	public void addTimeMeasurement(Long timeMeasurement) {
+		count.incrementAndGet();
+		totalTime.addAndGet(timeMeasurement);
+	}
+
+	public long getCount() {
+		return count.get();
+	}
+
+	public long getTotalTime() {
+		return TimeUnit.NANOSECONDS.toMillis(totalTime.get());
+	}
+
+	public ConcurrentHashMap<Class<?>, ModelUsageData> getSubModels() {
+		return subModels;
+	}
+
+	public double getAverageTime() {
+		return totalTime.get() / (double) count.get() / 1000000;
+	}
+
+	public void add(ModelUsageData statistics) {
+		this.count.addAndGet(statistics.count.get());
+		this.totalTime.addAndGet(statistics.totalTime.get());

```

Comment:
We are aware of the problem here. In our opinion it is not a significant one - we risk getting sightly invalid data when rendering the page with report. We can live with that for a while.
`AtomicReference` would be one solution to this. Please note, that we try to optimize performance of recording the stats events over correctness of the data being displayed to user.